custom_model_runner/datarobot_drum/drum/language_predictors/base_language_predictor.py

```diff
@@ -174,6 +197,94 @@ def _predict(self, **kwargs) -> RawPredictResponse:
+
+        if not is_streaming_response(response):
+            self._mlops_report_chat_prediction(
+                completion_create_params, start_time, response.choices[0].message.content
+            )
+            return response
+        else:
+
+            def generator():
+                message_content = ""
+                try:
+                    for chunk in response:
+                        message_content += (
+                            chunk.choices[0].delta.content
+                            if chunk.choices and chunk.choices[0].delta.content
+                            else ""
+                        )
+                        yield chunk
+                except Exception:
+                    self._mlops_report_error(start_time)
+                    raise
+
+                self._mlops_report_chat_prediction(
+                    completion_create_params, start_time, message_content
+                )
+
+            return generator()
+
+    def _chat(self, completion_create_params):
+        raise NotImplementedError("Chat is not implemented ")
+
+    def _mlops_report_chat_prediction(self, completion_create_params, start_time, message_content):
+        execution_time_ms = (time.time() - start_time) * 1000
+
+        self._mlops.report_deployment_stats(num_predictions=1, execution_time_ms=execution_time_ms)
+
+        latest_message = completion_create_params["messages"][-1]["content"]
+        features_df = pd.DataFrame([{"prompt": latest_message}])
+
+        predictions = [message_content]
+        try:
+            self._mlops.report_predictions_data(
+                features_df,
+                predictions,
+                association_ids=[str(uuid.uuid4())],
+                skip_drift_tracking=True,
+                skip_accuracy_tracking=True,
+            )
+        except DRCommonException:
+            logger.exception("Failed to report predictions data")
+
+    def _mlops_report_error(self, start_time):
+        execution_time_ms = (time.time() - start_time) * 1000
+
+        self._mlops.report_deployment_stats(num_predictions=0, execution_time_ms=execution_time_ms)
+
+    @staticmethod
+    def _validate_chat_response(response):
+        if getattr(response, "object", None) == "chat.completion":
+            return response
+
+        try:
+            # Take a peek at the first object in the iterable to make sure that this is indeed a chat completion chunk.
+            # This should catch cases where hook returns an iterable of a different type early on.
+            response_iter = iter(response)
+            first_chunk = next(response_iter)
+
+            if getattr(first_chunk, "object", None) == "chat.completion.chunk":
+                # Return a new iterable where the peeked object is included in the beginning
+                return itertools.chain([first_chunk], response_iter)
+        except StopIteration:
+            return iter(())
+        except Exception:
+            pass
```

Comment:
How about we log this somehow? who knows what kind of bug may be hiding here.

Reply:
Good point. I have changed this so that the new exception will be raised from the caught one which should preserve the context